texar/torch/run/executor.py

```diff
@@ -1483,6 +1493,37 @@ def erase_status_and_log_fn(executor: 'Executor'):
             self.test_log_format, self.test_progress_log_format,
             Event.TestingIteration, Event.Testing)
 
+    def _register_tbx_logging_actions(self):
+
+        # Register logging actions.
+        Points = Sequence[Union[Condition, Event]]
+        LogFn = Callable[['Executor'], str]
+
+        def _register(points: Points, fn: ActionFn):
+            for cond_or_event in points:
+                if isinstance(cond_or_event, Condition):
+                    self._register_action(cond_or_event, fn)
+                else:
+                    self._register_hook((cond_or_event, True), fn)
+
+        def tbx_train_log_fn(executor: 'Executor'):
+            train_metrics = executor.train_metrics
+
+            # log the metrics here
+            for key, value in train_metrics.items():
+                self.summary_writer.add_scalar(
+                    f"train/{key}", value.value(), executor.status["iteration"])
+
+        def tbx_eval_log_fn(executor: 'Executor'):
+            valid_metrics = executor.valid_metrics
+
+            for key, value in valid_metrics.items():
+                self.summary_writer.add_scalar(
+                    f"eval/{key}", value.value(), executor.status["epoch"])
```

Comment:
A potential problem here is `metric.value()` will be called twice after each validation (once for logging, once for tbX). For `StreamingMetric` this is not a problem, but for `SimpleMetric` this may be time consuming. Maybe we can modify the interface for `SimpleMetric` a bit, e.g., we can cache the previous result calculated in `value()`, until new values are added or `reset()` is called. This caching routine could be performed in the `SimpleMetric` class `value()` method, and subclasses of `SimpleMetric` would then need to implement another method (e.g. `_value()`) instead of `value()`.

Basically:

```python
class SimpleMetric(Metric[Input, Value]):
  _cached_value: Optional[Value]

  def reset(self):
    self._cached_value = None

  def add(...):
    self._cached_value = None

  def value(self):
    if self._cached_value is not None:
      return self._cached_value
    self._cached_value = self._value()
    return self._cached_value

  def _value(self) -> Value:
    r""" What previously goes in `value()` now goes here. """
    raise NotImplementedError
```

Reply:
Fixed in bbaf278